modules/mcc/include/opencv2/mcc/color.hpp

```diff
@@ -0,0 +1,354 @@
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+// Author: Longbu Wang <wanglongbu@huawei.com.com>
+//         Jinheng Zhang <zhangjinheng1@huawei.com>
+//         Chenqi Shan <shanchenqi@huawei.com>
+
+#ifndef __OPENCV_MCC_COLOR_HPP__
+#define __OPENCV_MCC_COLOR_HPP__
+
+#include <map>
+#include "opencv2/mcc/colorspace.hpp"
+#include "opencv2/mcc/distance.hpp"
+
+namespace cv
+{
+namespace ccm
+{
+
+/** @brief Color defined by color_values and color space
+*/
+
+class CV_EXPORTS_W Color
+{
+public:
+
+    /** @param grays mask of grayscale color
+        @param colored mask of colored color
+        @param history storage of historical conversion
+    */
+    Mat colors;
+    const ColorSpace& cs;
+    Mat grays;
+    Mat colored;
+    std::map<ColorSpace, std::shared_ptr<Color>> history;
+    Color(Mat colors_, enum COLOR_SPACE cs_) ;
+    Color(Mat colors_, enum COLOR_SPACE cs_, Mat colored_) ;
+    Color(Mat colors_, const ColorSpace& cs_, Mat colored_);
+    Color(Mat colors_, const ColorSpace& cs_);
+
+    virtual ~Color() {};
+
+    /** @brief Change to other color space.
+                 The conversion process incorporates linear transformations to speed up.
+        @param other type of ColorSpace.
+        @param  method the chromatic adapation method.
+        @param save when save if True, get data from history first.
+        @return Color.
+    */
+    Color to(COLOR_SPACE other, CAM method = BRADFORD, bool save = true);
+    Color to(const ColorSpace& other, CAM method = BRADFORD, bool save = true);
+    /** @brief Channels split.
+       @return each channel.
+    */
+    Mat channel(Mat m, int i);
+
+    /** @brief To Gray.
+    */
+    Mat toGray(IO io, CAM method = BRADFORD, bool save = true);
+
+    /** @brief To Luminant.
+    */
+    Mat toLuminant(IO io, CAM method = BRADFORD, bool save = true);
+
+    /** @brief Diff without IO.
+        @param other type of Color.
+        @param method type of distance.
+        @return distance between self and other
+    */
+    Mat diff(Color& other, DISTANCE_TYPE method = CIE2000);
+
+    /** @brief Diff with IO.
+        @param other type of Color.
+        @param io type of IO.
+        @param method type of distance.
+        @return distance between self and other
+    */
+    Mat diff(Color& other, IO io, DISTANCE_TYPE method = CIE2000);
+
+    /** @brief Calculate gray mask.
+    */
+    void getGray(double JDN = 2.0);
+
+    /** @brief Operator for mask copy.
+    */
+    Color operator[](Mat mask);
+
+};
+
+
+/** @brief Data is from https://www.imatest.com/wp-content/uploads/2011/11/Lab-data-Iluminate-D65-D50-spectro.xls
+           see Miscellaneous.md for details.
+*/
+const Mat ColorChecker2005_LAB_D50_2 = (Mat_<Vec3d>(24, 1) <<
```

Comment:
> ` << `

Avoid using of `MatCommaInitializer`. It significantly bloats binary code.
Use Mat wrapper over data from C arrays instead.

---

Global variables initialization must be removed from public headers.

---

Consider using lazy on-demand initialization of global static variables, see `getPredefinedDictionary(PREDEFINED_DICTIONARY_NAME name)` from `aruco/src/dictionary.cpp`.

Reply:
Done. This problem is solved by using array  to initialize Mat .

Reply:
> Global variables initialization must be removed from public headers.

Not yet.
Users don't need this information to use provided algorithms. It should be moved into src/.cpp or src/.hpp internal files.

Reply:
Thanks, we have moved color.hpp to src/color.cpp .